Force synced KJT to trace unbacked SymInt

ezyang opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #107788

Signed-off-by: Edward Z. Yang <ezyang@meta.com>

cc @voznesenskym @penguinwu @anijain2305 @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @Xia-Weiwen @wenzhe-nrv @jiayisunx @chenyang78 @aakhundov @kadeng

ezyang committed:
Force synced KJT to trace unbacked SymInt

Signed-off-by: Edward Z. Yang <ezyang@meta.com>

[ghstack-poisoned]

ezyang committed:
Update on "Force synced KJT to trace unbacked SymInt"


Signed-off-by: Edward Z. Yang <ezyangmeta.com>

cc voznesenskym penguinwu anijain2305 EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx chenyang78 aakhundov

[ghstack-poisoned]

voznesenskym commented on torch/_dynamo/variables/user_defined.py:
```diff
@@ -581,4 +581,9 @@ def __init__(self, value, **kwargs):
+    def var_getattr(self, tx, name):
+        source = AttrSource(self.source, name) if self.source else None
+        if source is not None and name in ("_length_per_key", "_offset_per_key"):
+            with torch._dynamo.config.patch(force_unspec_int_unbacked_size_like=True):
```
context manager hacky patching on config? broke
context manager hacky patching on instruction translator? woke

Have you seen `with tx.strict_translation_mode`? This feels like a sibling, `with tx.force_blah` - and you can then avoid internal settings on config. 

Otherwise, its not that uncommon to patch into a state like this and then resume tracing (down into an inline, up into a super, etc), so I would be okay with this in spirit :)

ezyang replied:
I also need a separate strategy for when we trace through a cumsum().unbind() call internally, in which case I either have to update torchrec to assert "these are sizes" or slide in those asserts myself... maybe just need to overwrite call_method appropriately or something

ezyang replied:
So I took a look at doing this, and one problem with putting it on `tx` is that we generate inner tx when we inline. But I want this property to inherit through the inner tx's. So this seems like, morally, not quite the right spot to put the knob.

ezyang replied:
I put it in TracingContext instead.

ezyang committed:
Update on "Force synced KJT to trace unbacked SymInt"


Signed-off-by: Edward Z. Yang <ezyangmeta.com>

cc voznesenskym penguinwu anijain2305 EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx chenyang78 aakhundov

[ghstack-poisoned]

suo commented:
@suo has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D48786784).

tugsbayasgalan commented on torch/_dynamo/variables/builder.py:
```diff
@@ -1039,11 +1040,21 @@ def wrap_unspecialized_primitive(self, value):
+                value, int
+            ):
+                wrapped_value = shape_env.create_unbacked_symint()
+                _constrain_range_for_size(wrapped_value)
```
nit: official API

ezyang committed:
Update on "Force synced KJT to trace unbacked SymInt"


Signed-off-by: Edward Z. Yang <ezyangmeta.com>

cc voznesenskym penguinwu anijain2305 EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx chenyang78 aakhundov

Differential Revision: [D48786784](https://our.internmc.facebook.com/intern/diff/D48786784)

[ghstack-poisoned]

ezyang committed:
Update on "Force synced KJT to trace unbacked SymInt"


Signed-off-by: Edward Z. Yang <ezyangmeta.com>

cc voznesenskym penguinwu anijain2305 EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx chenyang78 aakhundov

Differential Revision: [D48786784](https://our.internmc.facebook.com/intern/diff/D48786784)

[ghstack-poisoned]

ezyang committed:
Update on "Force synced KJT to trace unbacked SymInt"


Signed-off-by: Edward Z. Yang <ezyangmeta.com>

cc voznesenskym penguinwu anijain2305 EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx chenyang78 aakhundov

Differential Revision: [D48786784](https://our.internmc.facebook.com/intern/diff/D48786784)

[ghstack-poisoned]

ezyang commented on torch/_dynamo/variables/user_defined.py:
```diff
@@ -589,4 +590,9 @@ def __init__(self, value, **kwargs):
 
-    # TODO Handle getattr for _length_per_key and _offset_per_key properly.
+    def var_getattr(self, tx, name):
+        source = AttrSource(self.source, name) if self.source else None
+        if source is not None and name in ("_length_per_key", "_offset_per_key"):
```
oopsies

ezyang committed:
Update on "Force synced KJT to trace unbacked SymInt"


Signed-off-by: Edward Z. Yang <ezyangmeta.com>

cc voznesenskym penguinwu anijain2305 EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx chenyang78 aakhundov

Differential Revision: [D48786784](https://our.internmc.facebook.com/intern/diff/D48786784)

[ghstack-poisoned]

ezyang commented:
@pytorchbot merge

ezyang committed:
Update on "Force synced KJT to trace unbacked SymInt"


Signed-off-by: Edward Z. Yang <ezyangmeta.com>

cc voznesenskym penguinwu anijain2305 EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx chenyang78 aakhundov kadeng

Differential Revision: [D48786784](https://our.internmc.facebook.com/intern/diff/D48786784)

[ghstack-poisoned]

ezyang committed:
Update on "Force synced KJT to trace unbacked SymInt"


Signed-off-by: Edward Z. Yang <ezyangmeta.com>

cc voznesenskym penguinwu anijain2305 EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx chenyang78 aakhundov kadeng

Differential Revision: [D48786784](https://our.internmc.facebook.com/intern/diff/D48786784)

[ghstack-poisoned]

ezyang commented:
@pytorchbot merge

ezyang commented:
@pytorchbot revert -c weird -m "will test some pt2 models first"

ezyang commented:
@pytorchmergebot revert -c weird -m "reverting"

huydhn commented:
@pytorchbot drci

(testing the bot)

colin2328 commented on test/dynamo/test_torchrec.py:
```diff
@@ -0,0 +1,204 @@
+from torch.testing._internal.common_utils import NoTest
+
+try:
+    from torchrec.datasets.random import RandomRecDataset
```
in automated testing suite, my assumption is torchrec is rarely (never) available?

ezyang replied:
In previous PRs are made torchrec get installed for benchmark runs, but on further reflection I probably need to get it installed for test runs too

colin2328 replied:
@ezyang  sg.

 If this is something we want to test automatically, its more motivation to get prebuilt fbgemm (potentially through nova) cc @osalpekar , WDYT?

ezyang replied:
pre-built fbgemm would be quite profitable but difficult, as source changes to pytorch will invalidate fbgemm binaries

colin2328 reviewed (commented):
@ezyang  can you add more info to the description about what this diff does?

ezyang commented:
@pytorchbot revert -c weird -m "will test some pt2 models first"

huydhn commented:
@pytorchbot drci

(The bot doesn't like this PR somehow, it starts to work on other PR, but still not this one)

ezyang commented:
@pytorchbot revert -c nosignal -m weird

huydhn commented:
@pytorchbot drci
